pbench_logparse/main.go

```diff
@@ -0,0 +1,319 @@
+	if string(resultDir[len(resultDir)-1]) != "/" {
+		resultDir = resultDir + "/"
+	}
+
+	// If no block devices were passed, don't add to search
+	if len(blockDevices) > 0 {
+		fileHeader["disk_IOPS.csv"] = []string(blockDevices)
+	}
+
+	// If no network devices were passed, don't add to search
+	if len(netDevices) > 0 {
+		fileHeader["network_l2_network_packets_sec.csv"] = []string(netDevices)
+		fileHeader["network_l2_network_Mbits_sec.csv"] = []string(netDevices)
+	}
+
+	processList = strings.Split(processes, ",")
+	// If no process names were passed, don't add to search
+	if len(processList) > 0 {
+		fileHeader["cpu_usage_percent_cpu.csv"] = processList
+		fileHeader["memory_usage_resident_set_size.csv"] = processList
+	}
+
+	// Return director listing of searchDir
+	dirList, err := ioutil.ReadDir(searchDir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Iterate over directory contents
+	for _, item := range dirList {
+		// Match subdirectory that follows our pattern
+		if hostRegex.MatchString(item.Name()) && item.IsDir() {
+			kind := strings.Split(item.Name(), ":")
+			newHost := host{
+				kind:      kind[0],
+				resultDir: searchDir + item.Name(),
+			}
+			hosts = append(hosts, newHost)
+		}
+	}
+
+	// Maps are not ordered, create ordered slice of keys
+	keys := []string{}
+	for k := range fileHeader {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	// Iterate over all known hosts
+	for i, host := range hosts {
+		fmt.Printf("Host: %+v\n", host)
+		// Find each raw data CSV
+		for _, key := range keys {
+			fmt.Println(key)
+			fileList := findFile(host.resultDir, key)
+			// findFile returns slice, though there should only be one file
+			for _, file := range fileList {
+				// Parse file into 2d-string slice
+				result, err := readCSV(file)
+				if err != nil {
+					fmt.Printf("Error reading %v: %v\n", file, err)
+					continue
+				}
+				// In a single file we have multiple headers to extract
+				for _, header := range fileHeader[key] {
+					// Extract single column of data that we want
+					newResult, err := newSlice(result, header)
+					if err != nil {
+						//need to keep list of columns same for all types
+						//continue
+						fmt.Printf("newSlice returned error: %v\n", err)
+					}
+
+					// Mutate host to add calcuated stats to object
+					hosts[i].addResult(newResult, file, header)
+
+				}
+				fmt.Printf("CALLER Host populated: %+v\n", hosts[i])
+
+			}
+		}
+	}
+
+	err = writeCSV(keys, hosts)
+	if err != nil {
+		fmt.Printf("Error writing CSV: %v", err)
+	}
+}
+
+func writeCSV(keys []string, hosts []host) error {
+	csvFile, err := os.Create(resultDir + "out.csv")
+	if err != nil {
+		return err
+	}
+	defer csvFile.Close()
+
+	// Write test CSV data to stdout
+	writer := csv.NewWriter(csvFile)
+	defer writer.Flush()
+
+	// Create header & write
+	header := createHeaders(keys)
+	for _, h := range header {
+		writer.Write(h)
+	}
+
+	// TODO: Maybe use reflection to get these fields instead
```

Comment:
Yup, sorry about the "median", "mean" confusion. It does 2 things
- Reduces computation time as you told earlier.
- Increases flexibility for end-user.